openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+    MinMaxKernel(MinMaxKernel& other, tbb::split) : mMin(other.mMin), mMax(other.mMax) {}
+    math::MinMax<ValueT> run(const GridT &grid)
+    {
+        LeafMgr mgr(grid.tree());// super fast
+        tbb::parallel_reduce(mgr.leafRange(), *this);
+        return math::MinMax<ValueT>(mMin, mMax);
+    }
+    void operator()(const LeafRange& r)
+    {
+        for (auto leafIter = r.begin(); leafIter; ++leafIter) {
+            for (auto voxelIter = leafIter->beginValueOn(); voxelIter; ++voxelIter) {
+                const ValueT v = *voxelIter;
+                if (v < mMin) mMin = v;
+                if (v > mMax) mMax = v;
+            }
+        }
+    }
+    void join(const MinMaxKernel& other)
+    {
+        if (other.mMin < mMin) mMin = other.mMin;
+        if (other.mMax > mMax) mMax = other.mMax;
+    }
+    ValueT mMin, mMax;
+};//MinMaxKernel
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// Private class of FastSweeping to perform multi-threaded initialization
+template <typename GridT>
+struct FastSweeping<GridT>::DilateKernel
+{
+    using LeafRange = typename tree::LeafManager<TreeT>::LeafRange;
+    using BufferT = typename FastSweeping::CoordArrayT::ValueBuffer;//util::PagedArray<Coord, 10>::ValueBuffer;
+    using PoolT = tbb::enumerable_thread_specific<BufferT>;
+    DilateKernel(FastSweeping &parent)
+        : mParent(&parent), mPool(nullptr), mBackground(parent.mGrid1->background())
+    {
+    }
+    DilateKernel(const DilateKernel &parent) = default;// for tbb::parallel_for
+    DilateKernel& operator=(const DilateKernel&) = delete;
+
+    void run(int dilation, NearestNeighbors nn)
+    {
+#ifdef BENCHMARK_FAST_SWEEPING
+        util::CpuTimer timer("Construct LeafManager");
+#endif
+        tree::LeafManager<TreeT> mgr(mParent->mGrid1->tree());// super fast
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Changing background value");
+#endif
+        static const ValueT Unknown = std::numeric_limits<ValueT>::max();
+        changeLevelSetBackground(mgr, Unknown);//multi-threaded
+
+ #ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Dilating and updating mgr (parallel)");
+        //timer.restart("Dilating and updating mgr (serial)");
+#endif
+
+        const int delta = 5;
+        for (int i=0, d = dilation/delta; i<d; ++i) dilateActiveValues(mgr, delta, nn, IGNORE_TILES);
+        dilateActiveValues(mgr, dilation % delta, nn, IGNORE_TILES);
+        //for (int i=0, n=5, d=dilation/n; i<d; ++i) dilateActiveValues(mgr, n, nn, IGNORE_TILES);
+        //dilateVoxels(mgr, dilation, nn);
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Initializing grid and coords");
+#endif
+        BufferT prototype(mParent->mPagedArray);//exemplar used for initialization
+        PoolT pool(prototype);//thread local storage pool of ValueBuffers
+        mPool = &pool;
+        tbb::parallel_for(mgr.leafRange(32), *this);//multi-threaded
```

Comment:
perhaps use `tbb::task_scheduler_init::default_num_threads()` ?

Reply:
Ah actually I see you've used `std::thread::hardware_concurrency()` in some cases - I think would be fine instead of the tbb implementation too?

Reply:
Ideally it should scale with threads in your group in tbb; not based on core counts.  Might be running in a -j4 environment, for example.

Reply:
Think this still needs to be fixed in a number of placed

Reply:
32 is the grain-size, i.e. the smallest number of leaf nodes to be part of a tbb task. it's not the thread-point. 32 has proved to perform well on different platforms and certainly much better than the default value of 1. In other words I'm keeping it